src/Components/Web/src/Virtualization/VirtualizeBase.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides common functionality for virtualized lists.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public abstract class VirtualizeBase<TItem> : ComponentBase, IVirtualizeJsCallbacks, IAsyncDisposable
+    {
+        private VirtualizeJsInterop _jsInterop = default!;
+
+        private ElementReference _spacerBefore;
+
+        private ElementReference _spacerAfter;
+
+        [Inject]
+        private IJSRuntime JSRuntime { get; set; } = default!;
+
+        /// <summary>
+        /// Gets the number of items rendered before the visible part of the container.
+        /// </summary>
+        protected int ItemsBefore { get; private set; }
+
+        /// <summary>
+        /// Gets the number of items the container can visibly fit.
+        /// </summary>
+        protected int VisibleItemCapacity { get; private set; }
+
+        /// <summary>
+        /// Gets the number of items currently visible within the container.
+        /// </summary>
+        protected int ItemsVisible { get; private set; }
+
+        /// <summary>
+        /// Gets the total number of items.
+        /// </summary>
+        protected abstract int ItemCount { get; }
+
+        /// <summary>
+        /// Gets the size of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Renders the visible items in the list.
+        /// </summary>
+        /// <param name="builder">A <see cref="RenderTreeBuilder"/> to receive the render output.</param>
+        protected abstract void RenderItems(RenderTreeBuilder builder);
```

Comment:
This part of the API contract makes me not 100% convinced that `VirtualizeBase` should exist as an independent thing (instead of being merged with `VirtualizeDeferred`).

The reason is that `RenderItems(builder)` doesn't really communicate anything precise about the requirements for what this method should do. How would someone subclassing `VirtualizeBase` know what to do inside `RenderItems` to make it behave properly and virtualize things well? Most like the answer is "they shouldn't really inherit VirtualizeBase directly", in which case it would be great if it wasn't exposed as a thing you could subclass directly.

Perhaps it exists because of the split between `VirtualizeDeferred` and `VirtualizeFixed`. That gives an extra reason to be interested in possibly merging all three of them into a single thing (e.g., called `Virtualize`) if that's at all possible.

Sorry if it seems like we're going in circles on this question about which set of base classes should exist. If you think I'm giving contradictory suggestions please let me know! (It's totally possible, I do that sometimes.) Right now it feels like your `InfiniteScroll` component makes sense as it is, and the `VirtualizeDeferred` makes sense, but to me the other two classes (`VirtualizeFixed` and `VirtualizeBase`) seem tenuous.